packages/xstate-wallet/src/workflows/approve-budget-and-fund.ts

```diff
@@ -1,153 +1,343 @@
-  updateBudget: Action<WorkflowContext, any>;
+  hideUi: Action<Context, any>;
+  displayUi: Action<Context, any>;
+  sendResponse: Action<Context, any>;
+  sendBudgetUpdated: Action<Context, any>;
 }
-export type StateValue = keyof WorkflowStateSchema['states'];
-
-const generateConfig = (
-  actions: WorkflowActions
-): MachineConfig<WorkflowContext, WorkflowStateSchema, WorkflowEvent> => ({
-  id: 'approve-budget-and-fund',
-  initial: 'fundLedger',
-  entry: [actions.displayUi, actions.updateBudget],
-  states: {
-    waitForUserApproval: {
-      on: {
-        USER_APPROVES_BUDGET: {target: 'fundLedger', actions: []},
-        USER_REJECTS_BUDGET: {target: 'failure'}
-      }
-    },
-    fundLedger: {invoke: {src: 'createAndFundLedger', onDone: 'done'}},
-    done: {
-      type: 'final',
-      entry: [
-        actions.hideUi,
-        actions.sendResponse,
-        /* This might be overkill */ actions.sendBudgetUpdated
-      ]
-    },
-    failure: {type: 'final'}
-  }
-});
+export type StateValue = keyof Schema['states'];
 
-const mockActions: WorkflowActions = {
-  hideUi: 'hideUi',
-  displayUi: 'displayUi',
-  sendResponse: 'sendResponse',
-  sendBudgetUpdated: 'sendBudgetUpdated',
-  updateBudget: 'updateBudget'
-};
-
-export const approveBudgetAndFundWorkflow = (
+export const machine = (
   store: Store,
   messagingService: MessagingServiceInterface,
-  context: WorkflowContext
-): WorkflowMachine => {
-  const services: WorkflowServices = {
-    createAndFundLedger: (context: WorkflowContext) =>
-      CreateAndFundLedger.createAndFundLedgerWorkflow(store, {
-        initialOutcome: convertBudgetToAllocation(context),
-        participants: [context.player, context.hub]
-      })
+  context: Initial
+): StateMachine<Context, Schema, Event, Typestate> =>
+  createMachine<Context, Event, Typestate>({
+    id: 'approve-budget-and-fund',
+    context,
+    initial: 'waitForUserApproval',
+    entry: displayUI(messagingService),
+    states: {
+      waitForUserApproval: {
+        on: {
+          USER_APPROVES_BUDGET: {target: 'createLedger'},
+          USER_REJECTS_BUDGET: {target: 'failure'}
+        }
+      },
+      createLedger: {
+        invoke: {
+          id: 'createLedgerStartState',
+          src: initializeLedger(store),
+          onDone: {target: 'waitForPreFS', actions: setLedgerInfo}
+        }
+      },
+      waitForPreFS: {
+        invoke: {
+          id: 'subscribeToLedgerUpdates',
+          src: notifyWhenPreFSSupported(store),
+          onDone: {target: 'deposit', actions: assignDepositingInfo}
+        }
+      },
+      deposit: {
+        initial: 'init',
+        invoke: {
+          id: 'observeChain',
+          src: observeLedgerOnChainBalance(store)
+        },
+        on: {
+          CHAIN_EVENT: [
+            {target: 'done', actions: assignChainData, cond: fullAmountConfirmed},
+            {target: '.waitFullyFunded', actions: assignChainData, cond: myAmountConfirmed}
+          ]
```

Comment:
I did consider this option, but thought that it might be clearer to have the workflow managing the conditions through the conditions mechanism, keeping the service simpler.